File: jambeth/jambeth-persistence/src/main/java/de/osthus/ambeth/persistence/parallel/ParallelRunnable.java
package de.osthus.ambeth.persistence.parallel;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.locks.Lock;

import de.osthus.ambeth.collections.IList;
import de.osthus.ambeth.ioc.threadlocal.IForkState;
import de.osthus.ambeth.threading.IBackgroundWorkerParamDelegate;
import de.osthus.ambeth.util.ParamHolder;

public class ParallelRunnable<V> implements Runnable
{
	protected final RunnableHandle<V> runnableHandle;

	protected final boolean buildThreadLocals;

	public ParallelRunnable(RunnableHandle<V> runnableHandle, boolean buildThreadLocals)
	{
		this.runnableHandle = runnableHandle;
		this.buildThreadLocals = buildThreadLocals;
	}

	@Override
	public void run()
	{
		try
		{
			final Lock parallelLock = runnableHandle.parallelLock;
			IList<V> items = runnableHandle.items;
			IForkState forkState = runnableHandle.forkState;
			ParamHolder<Throwable> exHolder = runnableHandle.exHolder;
			CountDownLatch latch = runnableHandle.latch;

			IBackgroundWorkerParamDelegate<V> run = new IBackgroundWorkerParamDelegate<V>()
			{
				@Override
				public void invoke(V state) throws Throwable
				{
					runnableHandle.run.invoke(state);
					writeParallelResult(parallelLock, state);
				}
			};

			while (true)
			{
				V item;
				parallelLock.lock();
				try
				{
					if (exHolder.getValue() != null)
					{
						// an uncatched error occurred somewhere
						return;
					}
					// pop the last item of the queue
					item = items.popLastElement();
				}
				finally
				{
					parallelLock.unlock();
				}
				if (item == null)
				{
					// queue finished
					return;
				}
				try
				{
					if (buildThreadLocals)
					{
						forkState.use(run, item);
					}
					else
					{
						run.invoke(item);
					}
				}
				catch (Throwable e)
				{
					parallelLock.lock();
					try
					{
						if (exHolder.getValue() == null)
						{
							exHolder.setValue(e);
						}
					}
					finally
					{
						parallelLock.unlock();
					}
				}
				finally
				{
					latch.countDown();
				}
			}
		}
		finally
		{
			if (buildThreadLocals)
			{
				runnableHandle.threadLocalCleanupController.cleanupThreadLocal();
			}
		}
	}

	protected void writeParallelResult(Lock parallelLock, V item)
	{
		// intended blank
	}
}

File: ambeth/Ambeth.IoC/ambeth/util/ParallelRunnable.cs
using De.Osthus.Ambeth.Ioc.Threadlocal;
using De.Osthus.Ambeth.Threading;
using System;
using System.Collections.Generic;
using System.Threading;

namespace De.Osthus.Ambeth.Util
{
#if !SILVERLIGHT
    public class ParallelRunnable<V> : Runnable
    {
        protected readonly RunnableHandle<V> runnableHandle;

        protected readonly bool buildThreadLocals;

        public ParallelRunnable(RunnableHandle<V> runnableHandle, bool buildThreadLocals)
        {
            this.runnableHandle = runnableHandle;
            this.buildThreadLocals = buildThreadLocals;
        }

        public void Run()
        {
            try
            {
                Thread currentThread = Thread.CurrentThread;
			    String oldName = currentThread.Name;
			    if (buildThreadLocals)
			    {
				    String name = runnableHandle.createdThread.Name;
				    currentThread.Name = name + " " + oldName;
			    }
			    try
			    {
				    Object parallelLock = runnableHandle.parallelLock;
				    IList<V> items = runnableHandle.items;
				    IForkState forkState = runnableHandle.forkState;
				    ParamHolder<Exception> exHolder = runnableHandle.exHolder;
				    CountDownLatch latch = runnableHandle.latch;
                    
				    while (true)
				    {
					    V item;
					    lock (parallelLock)
					    {
						    if (exHolder.Value != null)
						    {
							    // an uncatched error occurred somewhere
							    return;
						    }
						    // pop the last item of the queue
						    item = items[items.Count - 1];
                            items.RemoveAt(items.Count - 1);
					    }
					    if (item == null)
					    {
						    // queue finished
						    return;
					    }
					    try
					    {
						    if (buildThreadLocals)
						    {
							    forkState.Use(runnableHandle.run, item);
						    }
						    else
						    {
                                runnableHandle.run(item);
						    }
					    }
					    catch (Exception e)
					    {
						    lock (parallelLock)
						    {
							    if (exHolder.Value == null)
							    {
								    exHolder.Value = e;
							    }
						    }
					    }
					    finally
					    {
						    latch.CountDown();
					    }
				    }
			    }
			    finally
			    {
				    if (buildThreadLocals)
				    {
					    currentThread.Name = oldName;
				    }
			    }
            }
            finally
            {
                if (buildThreadLocals)
                {
                    runnableHandle.threadLocalCleanupController.CleanupThreadLocal();
                }
            }
        }
    }
#endif
}